File: python/physics/bounce/bench.py
"""Benchmark for ball update — outputs JSON for the report generator.

Run: python bench.py
"""

import json
import timeit
from physics import Ball, update

WARMUP = 10_000
RUNS   = 1_000_000


def main() -> None:
    ball = Ball(x=400.0, y=300.0, vx=4.0, vy=3.0)

    # Warm up the interpreter
    for _ in range(WARMUP):
        update(ball, 30, 800, 600)

    # Time RUNS iterations; timeit returns total seconds for number=RUNS
    elapsed = timeit.timeit(
        lambda: update(ball, 30, 800, 600),
        number=RUNS,
    )

    mean_ns = elapsed / RUNS * 1e9
    print(json.dumps({
        "lang":    "python",
        "example": "bounce",
        "mean_ns": round(mean_ns, 4),
        "n":       RUNS,
    }))


if __name__ == "__main__":
    main()

File: nodejs/physics/bounce/src/bench.ts
/**
 * @file   bench.ts
 * @brief  Benchmark for ball update — outputs JSON for the report generator.
 *
 * Run: npx ts-node src/bench.ts
 */

import { Ball, update } from "./physics";

const WARMUP = 100_000;
const RUNS   = 1_000_000;

const ball: Ball = { x: 400, y: 300, vx: 4, vy: 3 };

// Warm up V8's JIT
for (let i = 0; i < WARMUP; i++) update(ball, 30, 800, 600);

const t0 = performance.now();
for (let i = 0; i < RUNS; i++) update(ball, 30, 800, 600);
const t1 = performance.now();

const meanNs = ((t1 - t0) * 1e6) / RUNS;  // performance.now() is in ms

console.log(JSON.stringify({
    lang:    "nodejs",
    example: "bounce",
    mean_ns: parseFloat(meanNs.toFixed(4)),
    n:       RUNS,
}));
